include/git2/cred.h

```diff
@@ -0,0 +1,308 @@
+	 */
+	GIT_CREDTYPE_USERNAME = (1u << 5),
+
+	/**
+	 * An SSH key-based authentication request
+	 *
+	 * Allows credentials to be read from memory instead of files.
+	 * Note that because of differences in crypto backend support, it might
+	 * not be functional.
+	 *
+	 * @see git_cred_ssh_key_memory_new
+	 */
+	GIT_CREDTYPE_SSH_MEMORY = (1u << 6),
+} git_credtype_t;
+
+/**
+ * The base structure for all credential types
+ */
+typedef struct git_cred git_cred;
+
+typedef struct git_cred_userpass_plaintext git_cred_userpass_plaintext;
+
+/** Username-only credential information */
+typedef struct git_cred_username git_cred_username;
+
+/** A key for NTLM/Kerberos "default" credentials */
+typedef struct git_cred git_cred_default;
+
+/**
+ * A ssh key from disk
+ */
+typedef struct git_cred_ssh_key git_cred_ssh_key;
+
+/**
+ * Keyboard-interactive based ssh authentication
+ */
+typedef struct git_cred_ssh_interactive git_cred_ssh_interactive;
+
+/**
+ * A key with a custom signature function
+ */
+typedef struct git_cred_ssh_custom git_cred_ssh_custom;
+
+/**
+ * Credential acquisition callback.
+ *
+ * This callback is usually involved any time another system might need
+ * authentication. As such, you are expected to provide a valid git_cred
+ * object back, depending on allowed_types (a git_credtype_t bitmask).
+ *
+ * Note that most authentication details are your responsibility - this
+ * callback will be called until the authentication succeeds, or you report
+ * an error. As such, it's easy to get in a loop if you fail to stop providing
+ * the same incorrect credentials.
+ *
+ * @param cred The newly created credential object.
+ * @param url The resource for which we are demanding a credential.
+ * @param username_from_url The username that was embedded in a "user\@host"
+ *                          remote url, or NULL if not included.
+ * @param allowed_types A bitmask stating which cred types are OK to return.
+ * @param payload The payload provided when specifying this callback.
+ * @return 0 for success, < 0 to indicate an error, > 0 to indicate
+ *       no credential was acquired
+ */
+typedef int GIT_CALLBACK(git_cred_acquire_cb)(
+	git_cred **cred,
+	const char *url,
+	const char *username_from_url,
+	unsigned int allowed_types,
+	void *payload);
+
+/**
+ * Free a credential.
+ *
+ * This is only necessary if you own the object; that is, if you are a
+ * transport.
+ *
+ * @param cred the object to free
+ */
+GIT_EXTERN(void) git_cred_free(git_cred *cred);
+
+/**
+ * Check whether a credential object contains username information.
+ *
+ * @param cred object to check
+ * @return 1 if the credential object has non-NULL username, 0 otherwise
+ */
+GIT_EXTERN(int) git_cred_has_username(git_cred *cred);
+
+/**
+ * Return the username associated with a credential object.
+ *
+ * @param cred object to check
+ * @return the credential username, or NULL if not applicable
+ */
+GIT_EXTERN(const char *) git_cred_get_username(git_cred *cred);
```

Comment:
Well, I'd usually note that this is an atypical API name, as we almost never add the "get" verb. But in fact compilers will complain if there is a function and typedef with the same name:

```C
#include <stdio.h>

typedef struct git_cred_username {
    char *username;
} git_cred_username;

char *git_cred_username(const git_cred_username *c)
{
    return c->username;
}

int main(int argc, const char *argv[])
{
    git_cred_username c = { "foo" };
    printf("Username: %s\n", git_cred_username(&c));
    return 0;
}
```

Compiling this results in the following error:

```bash
$ gcc main.c
main.c:7:7: error: 'git_cred_username' redeclared as different kind of symbol
 char *git_cred_username(const git_cred_username *c)
       ^~~~~~~~~~~~~~~~~
main.c:5:3: note: previous declaration of 'git_cred_username' was here
 } git_cred_username;
   ^~~~~~~~~~~~~~~~~
main.c: In function 'main':
main.c:14:22: error: expected ';' before 'c'
     git_cred_username c = { "foo" };
                      ^~
                      ;
main.c:15:49: error: 'c' undeclared (first use in this function)
     printf("Username: %s\n", git_cred_username(&c));
                                                 ^
main.c:15:49: note: each undeclared identifier is reported only once for each function it appears in
```

So calling it `get_username` is in fact the best way forward, even though uncommon.